File: src/main/java/io/vproxy/vfx/ui/shapes/BrokenLine.java
package io.vproxy.vfx.ui.shapes;

import io.vproxy.vfx.entity.Point;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.paint.Color;

import java.util.ArrayList;
import java.util.List;

public class BrokenLine {
    private final List<VLine> lines = new ArrayList<>();
    private final Group group = new Group();

    public BrokenLine(double width, double... points) {
        this(width, makePoints(points));
    }

    private static List<Point> makePoints(double[] points) {
        if (points.length % 2 != 0) {
            throw new IllegalArgumentException("points length must be an even number");
        }
        var ret = new ArrayList<Point>();
        Double x = null;
        for (var p : points) {
            if (x == null) {
                x = p;
            } else {
                ret.add(new Point(x, p));
                x = null;
            }
        }
        return ret;
    }

    public BrokenLine(double width, List<Point> points) {
        if (points.size() < 2) {
            throw new IllegalArgumentException("too few points to make a line");
        }
        for (int i = 1; i < points.size(); i++) {
            var current = points.get(i);
            var previous = points.get(i - 1);

            var line = new VLine(width);
            lines.add(line);
            group.getChildren().add(line);

            line.setStart(previous.x, previous.y);
            line.setEnd(current.x, current.y);
        }
    }

    public void setStartStyle(EndpointStyle style) {
        lines.get(0).setStartStyle(style);
    }

    public void setEndStyle(EndpointStyle style) {
        lines.get(lines.size() - 1).setEndStyle(style);
    }

    public void setStroke(Color color) {
        for (var l : lines) {
            l.setStroke(color);
        }
    }

    public Node getNode() {
        return group;
    }
}

File: src-ts/src/ui/shapes/BrokenLine.ts
import { Group } from '../../javafx/Pane.js';
import type { Node } from '../../javafx/Node.js';
import { Point } from '../../entity/Point.js';
import type { Color } from '../../javafx/color.js';
import { VLine } from './VLine.js';
import { EndpointStyle } from './EndpointStyle.js';

function makePoints(points: number[]): Point[] {
  if (points.length % 2 !== 0) {
    throw new Error('points length must be an even number');
  }
  const ret: Point[] = [];
  let x: number | null = null;
  for (const p of points) {
    if (x === null) {
      x = p;
    } else {
      ret.push(new Point(x, p));
      x = null;
    }
  }
  return ret;
}

export class BrokenLine {
  private readonly lines: VLine[] = [];
  private readonly group: Group = new Group();

  constructor(width: number, ...rest: Array<number | Point>) {
    const points: Point[] = rest.length === 0
      ? []
      : typeof rest[0] === 'number'
        ? makePoints(rest as number[])
        : (rest as Point[]);
    this.init(width, points);
  }

  private init(width: number, points: Point[]): void {
    if (points.length < 2) {
      throw new Error('too few points to make a line');
    }
    for (let i = 1; i < points.length; i++) {
      const current = points[i]!;
      const previous = points[i - 1]!;

      const line = new VLine(width);
      this.lines.push(line);
      this.group.getChildren().add(line);

      line.setStart(previous.x, previous.y);
      line.setEnd(current.x, current.y);
    }
  }

  setStartStyle(style: EndpointStyle): void {
    this.lines[0]!.setStartStyle(style);
  }

  setEndStyle(style: EndpointStyle): void {
    this.lines[this.lines.length - 1]!.setEndStyle(style);
  }

  setStroke(color: Color): void {
    for (const l of this.lines) {
      l.setStroke(color);
    }
  }

  getNode(): Node {
    return this.group;
  }
}
